### PR TITLE
fix(chat): finalize orphaned messages so chat can't hang on "generating"

### LINKED ISSUE #399: Coach chat freezes / gets stuck on "generating"
Multiple users report the coach chat hanging indefinitely on "generating" — especially on vague/multi-step requests and when editing injuries/excluded movements. This is the most frequently reported issue and blocks core usage.

## Reports

> What do I do when my chat has been stuck at generating for a week
>
> — **casecij583** in #general · 2026-05-09 01:12 UTC
> [view in Discord](https://discord.com/channels/1482942052898574336/1482942053443698721/1502478250528739471)

> Hey there. So i opened an account, hooked up gemini, went to chat but the chat is just hanging anytime i write anything
>
> — **llama_of_doom** in #general · 2026-04-12 13:06 UTC
> [view in Discord](https://discord.com/channels/1482942052898574336/1482942053443698721/1492873645830438984)

> Oh nevermind. It seems to be hanging specifically when im talking about injuries and blacking out certain movements
>
> — **llama_of_doom** in #general · 2026-04-12 13:08 UTC
> [view in Discord](https://discord.com/channels/1482942052898574336/1482942053443698721/1492874079055642725)

> I have the same issue, the coach often gets stuck.
>
> — **bsharpcyclist** in #general · 2026-04-12 19:08 UTC
> [view in Discord](https://discord.com/channels/1482942052898574336/1482942053443698721/1492964725229752521)

> Yep, I’m having the same freezing issue over and over
>
> — **llama_of_doom** in #general · 2026-04-12 19:26 UTC
> [view in Discord](https://discord.com/channels/1482942052898574336/1482942053443698721/1492969175008809233)

> How do you unstuck him (or her)?
>
> — **tacres** in #general · 2026-04-12 19:48 UTC
> [view in Discord](https://discord.com/channels/1482942052898574336/1482942053443698721/1492974678669787186)

> Roni is connected and can answer some simple questions fast, but during multi-step tasks or even some basic follow-ups it frequently freezes for 10–15+ minutes and stops responding.
> 
> Observed issues:
> 
> - Long hangs on simple questions
> - Gemini timeout behavior
> - Inconsistent response speed
> - Sometimes completes ID lookup, then freezes on follow-up
> - Hard to tell whether backend task is running or stuck
> 
> Main question:
> 
> Is this a known issue with Gemini / tool execution / Tonal API chaining?
> 
> And is there a recommended workflow to avoid freezes when trying to do workout creation or Tonal push tasks?
>
> — **projectot1000** in #general · 2026-05-26 14:06 UTC
> [view in Discord](https://discord.com/channels/1482942052898574336/1482942053443698721/1508833591977705522)

## Context

**Reported by:** casecij583, llama_of_doom, bsharpcyclist, tacres, projectot1000 (7+ messages across Apr–May).

---
_Filed automatically from a 90-day Discord backlog scan._

### PR DESCRIPTION
**Pull request by JeffOtano**
## Root-cause analysis

Issue #399 ("coach chat freezes / stuck on `generating`") is the most-reported bug — multiple users report the chat hanging indefinitely, sometimes for **days/weeks**, especially on multi-step requests and when editing injuries/excluded movements.

The chat's "generating" indicator is driven by this (non-time-bounded) condition in `src/features/chat/ChatThread.tsx`:

```js
const lastIsAssistantWithoutText = lastMessage?.role === "assistant" && !lastMessage.text.trim();
const lastIsFailed = lastMessage?.status === "failed";
const isThinking = !lastIsFailed && (lastIsRecentUser || lastIsAssistantWithoutText);
```

An **empty assistant message left in a non-terminal (`pending`) state** makes `isThinking` true *forever* across reloads (only the user-message branch is time-bounded).

How a message gets orphaned in `pending`:

- A turn streams inside an action (`convex/chatProcessing.ts` → `streamWithRetry`). The in-process resilience layer (`convex/ai/resilienceCircuitBreaker.ts`, `convex/ai/resilienceReporting.ts`) **does** finalize the message on every retry/exhaustion/terminal error — **but only when it actually runs.**
- If the generating action is **killed** — the Convex **600s action hard cap**, an OOM (cf. #400), or a hung tool call that outlives the in-stream 180s `AbortController` (`convex/ai/resilience.ts:237`; there is no per-tool timeout) — none of that finalize code runs.
- `@convex-dev/agent`'s built-in stream timeout (`streams.timeoutStream`, 10 min) only marks the **streaming row** `aborted`; it never flips the **persisted message** to a terminal status.
- **No cron reaps orphaned messages** (verified against `convex/crons.ts`).

Net effect: the persisted assistant message stays `pending` with empty text indefinitely → perpetual "generating" spinner.

## Fix

Add a **durable per-turn watchdog**, mirroring the agent library's own per-stream timeout pattern:

- New `convex/ai/stuckMessageWatchdog.ts`: `finalizeStuckMessagesForThread` internal mutation finalizes any message in a thread still non-terminal **past a grace window** to `{ status: "failed", error: "stuck_timeout" }`.
- `convex/chatProcessing.ts`: both turn entry points (`processMessage`, `continueAfterApproval` — the only callers of `streamWithRetry`) schedule the sweep up front via `ctx.scheduler.runAfter`. Scheduled functions **outlive the action that registered them**, so the sweep runs even when the action is killed.

Safety:
- `STUCK_MESSAGE_GRACE_MS = 10 min` equals the Convex action hard cap, so a message older than the cutoff can no longer have a live action working on it — **the sweep can never race an in-flight turn** (newer messages are skipped).
- `STUCK_MESSAGE_WATCHDOG_DELAY_MS = 15 min` > grace, so even a slow turn's assistant message has aged past the cutoff when the sweep fires.
- **No frontend change needed**: `ChatMessage.tsx:145` already renders a finalized `failed` empty assistant message as a retryable error alert ("Something went wrong generating a response… try again"). Finalizing to `failed` flips `lastIsFailed` true → spinner stops.

**No schema, auth, crypto, or webhook files touched.** No schema narrow.

## Test evidence (RED → GREEN)

New `convex/ai/stuckMessageWatchdog.test.ts` covers the reaper logic + edge cases.

**RED** (before the module existed — reproduces the absence of any reaper):
```
FAIL  convex/ai/stuckMessageWatchdog.test.ts [ convex/ai/stuckMessageWatchdog.test.ts ]
Error: Cannot find module './stuckMessageWatchdog' imported from .../stuckMessageWatchdog.test.ts
Test Files  1 failed (1)
```

**GREEN** (after fix):
```
Test Files  1 passed (1)
Tests  6 passed (6)
```

Cases: stuck `pending` message is failed; recent non-terminal message is left alone (in-flight turn); already-terminal messages ignored; stuck `streaming` status failed; only the stuck message (not a newer in-flight one in the same thread) is failed; delay > grace invariant.

## Local gates

| Gate | Result |
| --- | --- |
| `tsc --noEmit` | ✅ exit 0 |
| `eslint . --max-warnings=0` | ✅ exit 0 |
| `vitest run --coverage` | ✅ exit 0 (all suites pass, thresholds met) |
| `knip` | ✅ exit 0 |
| `prettier --check .` | ✅ all formatted |
| `scripts/check-file-size.sh` | ✅ exit 0 |

> Note: `npm run convex:smoke` was intentionally **not** run — no Convex deploy creds in this sandbox. No schema files changed.

## Open questions / unverified

- **Marked draft pending live verification.** The reaper logic is unit-tested with mocked component calls, but I could not exercise the full scheduled path against a live Convex deployment (no creds here). Worth confirming on dev that (a) `internal.ai.stuckMessageWatchdog.finalizeStuckMessagesForThread` resolves and runs, and (b) finalizing a `pending` message via `components.agent.messages.finalizeMessage` clears the `useUIMessages` "streaming" state as expected.
- `convex/_generated/api.d.ts` was hand-edited to register the new module (codegen needs deploy creds). Vercel's `convex deploy` will regenerate it on merge; the manual entry just keeps CI `tsc` green pre-merge.
- **Out of scope (follow-up):** this fixes the user-visible "stuck forever" symptom (the orphan reaper). It does **not** add a per-tool execution timeout, which is a likely *cause* of some 600s overruns/hangs. Happy to file a separate issue for bounding individual tool calls.

Closes #399

---
_Generated by [Claude Code](https://claude.ai/code/session_01VptyBZ72BZ3GPWszXsT3Ng)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Added automatic timeout handling for messages that remain stuck in a generating state. Such messages are now marked as failed after a grace period, preventing indefinite "generating" states.

### CODE REVIEW

File: convex/ai/stuckMessageWatchdog.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+// left non-terminal past the action lifetime so the UI can recover.
+
+// A message non-terminal longer than the Convex action hard cap can no longer
+// have a live action working on it, so failing it can never race a live turn.
+export const STUCK_MESSAGE_GRACE_MS = 10 * 60 * 1000;
+
+// Delay before the per-turn sweep runs. Exceeds the grace window plus the
+// worst-case context-build time so even a slow turn's assistant message has
+// aged past the grace cutoff by the time the sweep fires.
+export const STUCK_MESSAGE_WATCHDOG_DELAY_MS = 15 * 60 * 1000;
```
[the remark this answers is not in the document]
Comment (JeffOtano) [reply convex/ai/stuckMessageWatchdog.ts]: Good catch — this was a real completeness bug, fixed in 49ec6d2.

The watchdog timings are now derived from the Convex action cap (`CONVEX_ACTION_MAX_MS = 10m`):
- `STUCK_MESSAGE_GRACE_MS = cap + 1m` (a swept row's owning action is provably dead, since it started no later than the row's creation).
- `STUCK_MESSAGE_WATCHDOG_DELAY_MS = grace + cap + 1m` (= 22m), so even a final-attempt row created at the very end of the action's life has aged past the grace window when the single per-turn sweep fires.

Added a regression test asserting `WATCHDOG_DELAY - GRACE > CONVEX_ACTION_MAX_MS` so this invariant can't silently regress. Went with the larger fixed delay rather than self-rescheduling to keep the sweep stateless.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by JeffOtano** [comment]
Addressed review feedback in 49ec6d2:

- **Codex P2 (late-created row stranded):** fixed — watchdog delay now exceeds grace + the action cap so a final-attempt assistant row created late in the action's life is still swept. Regression test added.
- **CodeRabbit — magic `numItems: 50`:** extracted to `STUCK_MESSAGE_SWEEP_PAGE_SIZE`.
- **CodeRabbit — docstring coverage:** added a docstring to the scheduled mutation.
- **CodeRabbit — per-turn redundant sweeps (dedupe per thread):** intentionally declining. Deduping requires persisting a per-thread scheduled-job id and cancel/reschedule on every turn (new mutable state + a cancel/reschedule race) for negligible benefit — each sweep is one paginated read plus a finalize of only stale rows, and is a no-op on a healthy thread. Per AGENTS.md's simplicity/YAGNI guidance I'm keeping the sweep stateless; happy to revisit if scheduler volume ever shows up as a real cost.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by JeffOtano** [comment]
@codex review

**Comment by JeffOtano** [comment]
@coderabbitai review

(Re-review requested for the latest commit `49ec6d2`, which addresses Codex's P2 timing bug and the two CodeRabbit nitpicks. The prior approval was on `36fa949`.)

---
_Generated by [Claude Code](https://claude.ai/code)_
